extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/storage/sql/SQLCatalogManager.java

```diff
@@ -0,0 +1,578 @@
+      SAFETY_CHECK;
+
+  private static final String UPDATE_STATE =
+      UPDATE_HEAD +
+      "  state = :state,\n" +
+      "  updateTime = :updateTime\n" +
+      WHERE_TABLE_ID;
+
+  private static final String SELECT_TABLE =
+      "SELECT creationTime, updateTime, state, payload\n" +
+      "FROM %s\n" +
+      WHERE_TABLE_ID;
+
+  private static final String SELECT_PAYLOAD =
+      "SELECT state, payload\n" +
+      "FROM %s\n" +
+      WHERE_TABLE_ID;
+
+  private static final String SELECT_ALL_TABLES =
+      "SELECT schemaName, name\n" +
+      "FROM %s\n" +
+      "ORDER BY schemaName, name";
+
+  private static final String SELECT_TABLES_IN_SCHEMA =
+      "SELECT name\n" +
+      "FROM %s\n" +
+      "WHERE schemaName = :schemaName\n" +
+      "ORDER BY name";
+
+  private static final String SELECT_TABLE_DETAILS_IN_SCHEMA =
+      "SELECT name, creationTime, updateTime, state, payload\n" +
+      "FROM %s\n" +
+      "WHERE schemaName = :schemaName\n" +
+      "ORDER BY name";
+
+  private static final String DELETE_TABLE =
+      "DELETE FROM %s\n" +
+      WHERE_TABLE_ID;
+
+  private final MetastoreManager metastoreManager;
+  private final SQLMetadataConnector connector;
+  private final ObjectMapper jsonMapper;
+  private final IDBI dbi;
+  private final String tableName;
+  private final Deque<Listener> listeners = new ConcurrentLinkedDeque<>();
+
+  @Inject
+  public SQLCatalogManager(MetastoreManager metastoreManager)
+  {
+    if (!metastoreManager.isSql()) {
+      throw new ISE("SQLCatalogManager only works with SQL based metadata store at this time");
+    }
+    this.metastoreManager = metastoreManager;
+    this.connector = metastoreManager.sqlConnector();
+    this.dbi = connector.getDBI();
+    this.jsonMapper = metastoreManager.jsonMapper();
+    this.tableName = getTableDefnTable();
+  }
+
+  @Override
+  @LifecycleStart
+  public void start()
+  {
+    createTableDefnTable();
+  }
+
+  @Override
+  public void stop()
+  {
+  }
+
+  // Mimics what MetadataStorageTablesConfig should do.
+  public String getTableDefnTable()
+  {
+    final String base = metastoreManager.tablesConfig().getBase();
+    if (Strings.isNullOrEmpty(base)) {
+      return TABLES_TABLE;
+    } else {
+      return StringUtils.format("%s_%s", base, TABLES_TABLE);
+    }
+  }
+
+  // TODO: Move to SqlMetadataConnector
+  @Override
+  public void createTableDefnTable()
+  {
+    if (!metastoreManager.createTables()) {
+      return;
+    }
+    connector.createTable(
+        tableName,
+        ImmutableList.of(
+            StringUtils.format(
+                "CREATE TABLE %s (\n"
+                + "  schemaName VARCHAR(255) NOT NULL,\n"
+                + "  name VARCHAR(255) NOT NULL,\n"
+                + "  creationTime BIGINT NOT NULL,\n"
+                + "  updateTime BIGINT NOT NULL,\n"
+                + "  state CHAR(1) NOT NULL,\n"
+                + "  payload %s,\n"
+                + "  PRIMARY KEY(schemaName, name)\n"
```

Comment:
Does this need a IF NOT EXIST clause? This seems to be called on lifecycle start, which I believe is called whenever there is a leadership change to this service instance / the cluster comes up. Wont this fail if the table already exists?

Reply:
ahh. And yes I do see now that we catch the exception at deeper level, though as you said, we dont differentiate between the table existing and not being able to create table / other errors. Thanks for clarifying